### PR TITLE
Implement case mapping internals

### LINKED ISSUE #1226: Implement core case mapping functionality
Support for case folding is the biggest missing piece for my work with irregexp. Dumping my thoughts on how we should tackle it:

## Scope

ICU4C provides low-level character/string case mapping code in `common/ucase.h`. Implementing the equivalent functionality in ICU4X seems like a good first step. The CaseMap class in `common/unicode/casemap.h` could be implemented as a follow-up.

## Strategy
`common/ucase.cpp` defines a UCaseProps class, which contains a codepoint trie and several auxiliary data arrays. A single instance, `ucase_props_singleton`, is initialized based on the data in `common/ucase_props_data.h`, which is auto-generated by `tools/unicode/c/genprops/casepropsbuilder.cpp`. The contents of the codepoint trie are a 16-bit value with an assortment of flags and a payload that can either represent the signed delta between a codepoint and its corresponding upper/lowercase version, or an unsigned index into a sidetable of exception data.

The most straight-forward approach is to reproduce that data structure in Rust and expose roughly the same API. We can export the data from ICU4C using a similar approach to unicode properties.

## Survey

I've split up the functionality in `ucase.h` into several categories.

### Simple case folding
These functions are obvious inclusions:
- ucase_tolower / ucase_toupper / ucase_totitle: convert  a char to lower/upper/title case
- ucase_fold: simple case folding for a single char
- ucase_getType: return the type (lower/upper/title/none) of a single char
- ucase_isSoftDotted / ucase_isCaseSensitive: whether a character is soft-dotted/case-sensitive. These have dedicated bits in the trie value, so they're easy to implement.

### Full case folding
Full case folding includes codepoints that map to multiple codepoints and vice versa. This uses an additional sidetable, `unfold`. In theory we could expose a version of this code that does not support full case folding to save memory for users who don't need it; however, the sidetable is currently just a `[u16; 370]`, so it hardly seems worth it.
- ucase_toFullLower / ucase_toFullUpper / ucase_toFullTitle / ucase_toFullFolding: the full case-folding equivalents of the simple case-folding functions above
- ucase_getCaseLocale: This converts a locale string to an enum to be fed into locale-sensitive functions.
- UCaseContext / UCaseContextIterator: These are used to examine surrounding context in cases where that matters for casefolding.

### Case closure
Given a char/string and a set, these functions add all case-mapped equivalents of the input to the set. We will likely want to modify the API to make it more idiomatic.
- ucase_addCaseClosure
- ucase_addStringCaseClosure

### Questionable utility
- FullCaseFoldingIterator: This iterates over all codepoints with more than one codepoint in their full case folding. Not sure what the usecase is.
- ucase_addPropertyStarts: AFAICT, this is used for populating unicode sets. We currently precompute those sets instead.
- LatinCase::*: See question 1 below.

## Questions 
1. **Overlap with Rust std**: `std::char` already provides `to_uppercase()` and `to_lowercase()` (plus a variety of ascii-specific methods that I think make LatinCase obsolete), and it's awkward to duplicate functionality. However, it's not clear that we can build everything we need on top of Rust's stdlib. Furthermore, Rust updates to new versions of the Unicode standard regularly and [does not consider this a breaking change](https://doc.rust-lang.org/std/char/constant.UNICODE_VERSION.html), which could be a problem for users who expect their case mapping to be stable only to see it change when compiling with a more recent Rust compiler.
2. **Data representation**: The representation of data inside UCaseProps is not very Rust-idiomatic: for example, the exceptions data is an array of raw u16 values representing variable-length entries with a header and a variable number of slots. However, it seems space-efficient and fast to access, so it's likely best not to mess with it.

### PR DESCRIPTION
**Pull request by iainireland**
Fixes #1226, except for the `ucase_toFull(Upper|Lower|Title|Folding)` functions. Those functions require a context string to be passed in, and still need some design; since they aren't needed for my irregexp work, I'm splitting them off into a separate issue.

Other than those, this is a fairly direct re-implementation of ucase.cpp. Some implementation notes:
1. Following ucase_props_data.h, CaseMapping has three components: a code point trie that contains simple data covering most characters, an exception table for other characters, and an unfold map for case closure of strings.
2. I used the same data representation for the exception table as ICU4C: a packed array of variable-length entries, stored as an array of u16. The data is validated on construction. 
3. In ICU4C, the unfold map is stored (roughly speaking) as a sorted array of key/value pairs, where both the keys and the values are fixed-length UTF16 strings. Using the raw data directly in Rust turned out to be awkward because Rust is natively UTF8, so after a bit of benchmarking, I used a ZeroMap instead. (I actually benchmarked LiteMap vs a direct port vs ZeroMap: ZeroMap was the slowest, but LiteMap was the fastest, and they were all pretty close. I went with ZeroMap for memory reasons and in the hopes that it may be optimized better in the future now that we're actually using it.)
4. I put the provider code in the existing uprops provider, because it seems related, and because both providers need the ability to read code point tries.
5. Most of the tests are in the provider, where the test data is available. Some functions that are unlikely to be public in the final API are public here so that we can test them in another crate. (Of the currently implemented functions, I think only to_upper/to_lower/to_title are likely to be public in the final API.) Once we've written higher-level code that makes use of the public-for-testing functions, we can make them internal-only.
6. The SetAdder trait is a placeholder.
7. The data is from my in-flight PR for icuexportdata (https://github.com/unicode-org/icu/pull/1938).

### CODE REVIEW

File: provider/uprops/src/casemapping.rs
Comment (sffc) [inline provider/uprops/src/casemapping.rs:35]: Suggestion: It would be nice to make a utility function to convert from the TOML CPT Header to the ICU4X CPT Header.  The utility function could live in this crate for now (perhaps in a "helpers" module) and be shared between both this code and the uprops code.  I see that you already re-use `super::enumerated::EnumeratedPropertyCodePointTrie` in your serde definition, so this should be fairly easy to do.

File: experimental/casemapping/src/internals.rs
Comment (sffc) [inline experimental/casemapping/src/internals.rs:616]: Suggestion: I think this sort of decoding should be done in icuexportdata.  The TOML files we export should be "human readable" except for a small, specific set of exceptions, like CPTs.  If the data is really just a map from strings to strings, then the TOML should be a map of strings to strings.  Consumers of the TOML can readily massage the data into whatever format they like, including going back to an ICU4C-like structure.

If this is too much work, consider this suggestion optional.

File: provider/uprops/src/uprops_serde.rs
Comment (sffc) [inline provider/uprops/src/uprops_serde.rs:96]: Nit: Rename `EnumeratedPropertyCodePointTrie` to something more general.

File: experimental/casemapping/src/internals.rs
Comment (sffc) [inline experimental/casemapping/src/internals.rs:468]: Issue: No panics in runtime code (unless you can prove they are unreachable).

I assume this is runtime code, but if this is data builder code, you should hide it behind a flag such as `provider_transform_internals` like we do elsewhere.

File: experimental/casemapping/src/internals.rs
Comment (sffc) [inline experimental/casemapping/src/internals.rs:289]: Nit: Hide transformer-specific code behind the feature `provider_transform_internals`

**Review by sffc — DISMISSED** [review]

**Comment by iainireland** [comment]
@hsivonen Zibi thought this patch might be relevant to you.

File: provider/uprops/src/casemapping.rs
Comment (iainireland) [reply provider/uprops/src/casemapping.rs:35]: I went ahead and implemented TryFrom<&SerializedCodePointTrie> for CodePointTrieHeader.

File: experimental/casemapping/src/internals.rs
Comment (iainireland) [reply experimental/casemapping/src/internals.rs:616]: I'm going to skip this one. Given that the other two pieces of data in this file are a CPT (opaque) and a packed array of variable length data structures (extremely opaque), I don't think there's a lot of value in making this particular field human readable. I expect that the data provider on the ICU4X side will be the only consumer of this TOML.

File: provider/uprops/src/uprops_serde.rs
Comment (iainireland) [reply provider/uprops/src/uprops_serde.rs:96]: `uprops_serde::enumerated::EnumeratedPropertyCodePointTrie` -> `uprops_serde::SerializedCodePointTrie`

File: experimental/casemapping/src/internals.rs
Comment (iainireland) [reply experimental/casemapping/src/internals.rs:468]: (This is runtime code.)

The design here is that `try_from_icu` is the only way to build a CaseMapping. In that function, we call the same helper function on every full mapping string and propagate errors. It is therefore impossible (short of writing broken serialized data by hand and deserializing it) for this to fail.

I will change the expect message here to something like "Data validated on construction".

Comment (sffc) [reply experimental/casemapping/src/internals.rs]: `try_from_icu` runs in the data transformer, but the class could also be constructed from serialized postcard data, correct? If so, the current policy is no panicking for code reachable by malformed/malicious data.

Comment (sffc) [reply experimental/casemapping/src/internals.rs]: Related: https://github.com/unicode-org/icu4x/issues/1183

File: experimental/casemapping/src/internals.rs
Comment (sffc) [inline experimental/casemapping/src/internals.rs:734]: Nit: You still have some tabs here.  Surprised `rustfmt` doesn't fix it for you.

**Comment by sffc** [comment]
Also, please fix the merge conflict.

**Comment by iainireland** [comment]
This has been sitting long enough waiting for review that I'm mostly done the next chunk of work, which restructures some of the internals, so I'm going to convert this to a draft and we can review the whole thing at once.

**Comment by iainireland** [comment]
Okay, here's the reworked version. As we discussed in the deep dive, it includes internal support for various operations needed to implement case closure and titlecase, but doesn't expose them as part of the public API.

File: experimental/casemapping/src/casemapping.rs
Comment (sffc) [inline experimental/casemapping/src/casemapping.rs:28]: Nit:

```suggestion
    pub fn try_new<P>(provider: &P) -> Result<CaseMapping, DataError>
```

File: experimental/casemapping/src/casemapping.rs
Comment (sffc) [inline experimental/casemapping/src/casemapping.rs:36]: Nit:

```suggestion
    pub fn try_new_with_locale<P>(provider: &P, locale: &Locale) -> Result<CaseMapping, DataError>
```

File: experimental/casemapping/src/casemapping.rs
Comment (sffc) [inline experimental/casemapping/src/casemapping.rs:28]: Suggestion (optional): Consider changing the name to `try_new_default()` or `try_new_global()` or something similar to indicate the assumption that this method is making.  If you can't come up with a sufficiently good name then leave it as `try_new()`.

File: experimental/casemapping/src/casemapping.rs
Comment (sffc) [inline experimental/casemapping/src/casemapping.rs:76]: Here and elsewhere:

Praise: Thank you for stating this caveat in the docs.

Request: Please link to the method you should use if you want the full behavior.

File: experimental/casemapping/src/error.rs
Comment (sffc) [inline experimental/casemapping/src/error.rs:10]: Request: make this `Copy` if possible.  #1396

File: experimental/casemapping/src/error.rs
Comment (sffc) [inline experimental/casemapping/src/error.rs:25]: Nit:

```suggestion
    pub(crate) fn invalid<T>(reason: &'static str) -> Result<T, Self> {
```

File: experimental/casemapping/src/exceptions.rs
Comment (sffc) [inline experimental/casemapping/src/exceptions.rs:16]: Discussion: I've said this before verbally, but I'll put it in writing. It really would be a *lot* more elegant with regard to code readability, ease of validation, and consistency across ICU4X if you would make this `VarZeroVec<ExceptionItemULE>` with something along the lines of

```rust
#[derive(VarULE)]
struct ExceptionItemULE {
    header: /* this can be a structured header type */,
    replacements: str, // you can put your UTF-8 in here and elimiate the side table
}
```

I'm quite confident you could cut the size of this file, and probably some of the sister files, at least in half if you did it this way.

The *only* downside I'm aware of is that that this incurs an additional "hop" since we need to look at the index table.  I'd be interested to see if this is a big enough penalty to show up in benchmarks.

Another model that might solve the problem would be to store code point replacements in the side table along with string replacements.  You could then have `ZeroVec<ExceptionItemULE>` where `ExceptionItemULE` is a fixed-width type containing the indices into the string.  You then do not need any other lookups.

File: experimental/casemapping/src/provider.rs
Comment (sffc) [inline experimental/casemapping/src/provider.rs:18]: Nit: You can and should remove the `key` module now that the key is in the ResourceMarker.

File: experimental/casemapping/src/internals.rs
Comment (sffc) [reply experimental/casemapping/src/internals.rs]: (this is still unresolved and I'm unsure how to move forward)

File: experimental/casemapping/src/casemapping.rs
Comment (iainireland) [reply experimental/casemapping/src/casemapping.rs:28]: I was thinking of this along the same lines as `String`, although there it's `String::new` and String::with_capacity` (not `new_with_capacity`). I have a slight preference for leaving this as `try_new` and describing the default locale behaviour in the doc comment instead of trying to fit it in the name, but if there are other cases in ICU4X where we use something like "try_new_default" then consistency is probably enough of an argument to sway me.

File: experimental/casemapping/src/error.rs
Comment (iainireland) [reply experimental/casemapping/src/error.rs:10]: DecodeUTF16Error and CodePointTrieError are not `Copy`, and the latter is not yet `Clone`. Easy enough to make both of our own errors `Clone`, at least.

**Comment by iainireland** [comment]
@markusicu Review ping (hoping for any domain-specific comments you might have about case mapping)

File: experimental/casemapping/src/error.rs
Comment (sffc) [reply experimental/casemapping/src/error.rs:10]: Discussion: https://github.com/unicode-org/icu4x/issues/1396#issuecomment-1035509876

**Comment by iainireland** [comment]
@echeran Have you had a chance to look at this yet?

File: experimental/casemapping/src/exceptions_builder.rs
Comment (echeran) [inline experimental/casemapping/src/exceptions_builder.rs:10]: suggestion: add a comment about usage pattern for motivating use case

IIRC, this is intended to be used prior to runtime, when the data is being transformed. That contrasts with `UnicodeSetBuilder`, which is intended as a runtime thing. It might be good to document, since most `*Builder`s I've seen follow the fluent builder pattern for runtime user APIs.

File: experimental/casemapping/src/internals.rs
Comment (echeran) [inline experimental/casemapping/src/internals.rs:639]: suggestion: have named constants for chars, too

Could these chars have named consatants like lines 623-628 (more readable)?

File: experimental/casemapping/src/internals.rs
Comment (echeran) [inline experimental/casemapping/src/internals.rs:623]: suggestion(optional): use 4 hex digits for BMP -- `\u{XXXX}`

Only because these are code points, I'm only used to seeing code points described as U+XXXX with 4 or more hex digits.

FWIW, C++ either allows 4 or 8 hex digits in Unicode char escapes: https://en.cppreference.com/w/cpp/language/escape

File: provider/uprops/src/casemapping.rs
Comment (echeran) [inline provider/uprops/src/casemapping.rs:73]: [`uprops_toml_root()`](https://github.com/unicode-org/icu4x/blob/main/provider/testdata/src/paths.rs#L18)

File: provider/uprops/src/casemapping.rs
Comment (echeran) [inline provider/uprops/src/casemapping.rs:120]: [`uprops_toml_root()`](https://github.com/unicode-org/icu4x/blob/main/provider/testdata/src/paths.rs#L18)

File: experimental/casemapping/src/internals.rs
Comment (echeran) [inline experimental/casemapping/src/internals.rs:734]: suggestion: go ahead

Based on the comments in [ICU-13416](https://unicode-org.atlassian.net/browse/ICU-13416), it looks like this was given a lot of thought and done at the request of a real use case with approval from various commitees' members.

I like the idea of consistency with ICU, give the work is already there (thanks!).

File: experimental/casemapping/src/internals.rs
Comment (echeran) [inline experimental/casemapping/src/internals.rs:739]: question: would one of these need `CaseMapLocale::WesternArmenian`?

File: provider/uprops/src/casemapping.rs
Comment (echeran) [inline provider/uprops/src/casemapping.rs:64]: suggestion: include a few more more tests

The tests `StringCaseTest::TestGreekUpper` and `StringCaseTest::TestArmenian()` seem like short tests to implement. The latter depends on whether we match ICU on Eastern vs. Western Armenian.

Also, `StringCaseTest::TestLongUpper()` seems interesting and short. Maybe Rust makes that a non-issue, but nice to have.

All the other tests seem like they are about titlecasing, UTF-8 and/or Edits, so no other tests of interest.

File: provider/uprops/src/uprops_serde.rs
Comment (echeran) [inline provider/uprops/src/uprops_serde.rs:87]: praise: Thanks!

**Review by echeran — CHANGES_REQUESTED** [review]
Looks good -- a majority of things I commented are related to info clarity in explanatory source comments.

File: experimental/casemapping/Cargo.toml
Comment (iainireland) [reply experimental/casemapping/Cargo.toml:43]: We already use `provider_transform_internals` in a few other places (eg the datetime component). It's documented in components/icu/src/lib.rs, under Features.

File: experimental/casemapping/src/exceptions.rs
Comment (iainireland) [reply experimental/casemapping/src/exceptions.rs:13]: I'll polish documentation in a future patch.

File: experimental/casemapping/src/casemapping.rs
Comment (iainireland) [reply experimental/casemapping/src/casemapping.rs:106]: The case folding data in ICU defines four kinds of case folding: common (C), simple (S), full (F), and Turkic (T). 

See here: https://www.unicode.org/Public/12.1.0/ucd/CaseFolding.txt

I'll polish documentation in a future patch.

File: experimental/casemapping/src/exceptions.rs
Comment (iainireland) [reply experimental/casemapping/src/exceptions.rs:14]: Will fix in future patch.

File: experimental/casemapping/src/exceptions.rs
Comment (iainireland) [reply experimental/casemapping/src/exceptions.rs:73]: Will fix in future patch. I rewrite all this code anyway.

File: experimental/casemapping/src/exceptions.rs
Comment (iainireland) [reply experimental/casemapping/src/exceptions.rs:76]: Will fix in future patch.

File: experimental/casemapping/src/exceptions.rs
Comment (iainireland) [reply experimental/casemapping/src/exceptions.rs:130]: Fixed in future patch.

File: experimental/casemapping/src/exceptions.rs
Comment (iainireland) [reply experimental/casemapping/src/exceptions.rs:105]: This all gets rewritten in a future patch.

File: experimental/casemapping/src/exceptions_builder.rs
Comment (iainireland) [reply experimental/casemapping/src/exceptions_builder.rs:10]: Will fix in future patch.

File: experimental/casemapping/src/internals.rs
Comment (iainireland) [reply experimental/casemapping/src/internals.rs:639]: I mostly followed the ICU4C code here. I can take another pass and clean it up (in a future patch).

File: provider/uprops/src/casemapping.rs
Comment (iainireland) [reply provider/uprops/src/casemapping.rs:73]: Oh, cool, thanks!

File: experimental/casemapping/src/internals.rs
Comment (iainireland) [reply experimental/casemapping/src/internals.rs:734]: I'll push this off for a future patch to think about the question below.

File: experimental/casemapping/src/internals.rs
Comment (iainireland) [reply experimental/casemapping/src/internals.rs:739]: Good question, I'll look into it.

File: experimental/casemapping/src/internals.rs
Comment (iainireland) [reply experimental/casemapping/src/internals.rs:623]: Will consider for future patch.

File: provider/uprops/src/casemapping.rs
Comment (iainireland) [reply provider/uprops/src/casemapping.rs:64]: TestArmenian depends on enabling the Armenian code that's currently commented out.
TestGreekUpper tests a separate code path in ICU4C that I haven't ported yet.

Will make a note to revisit this in the future.

**Comment by iainireland** [comment]
Okay, finally got a chance to page this back in. I've taken notes on all the feedback about improved comments/documentation, fixed up a couple of places where Elango had concrete suggestions (eg `uprops_toml_root`), and rebased. I'll address the remaining feedback in a subsequent PR. 

If I can get a stamp, I think this is ready to land.

**Review by sffc — APPROVED** [review]
Approval Stamp

Note that Googlers are not working tomorrow (Friday) so if you need to get this merged before Monday you may need to manually dismiss Elango's "changes requested" review.

**Pull request merged**
